src/ir.rs

```diff
@@ -63,24 +88,112 @@ impl Cfg {
         }
 
         // build dominator tree
-        // TODO: actually build a reasonable dominator tree rather than this sad, totally naive approximation where every other block is a child of the entry block.
-        // build edges
-        for id in blocks.keys() {
-            let parent = if id.is_entry() {
-                None
-            } else {
-                Some(BlockId::ENTRY)
-            };
-            cfg.map.get_mut(id).unwrap().immediate_dominator = parent;
-        }
-        cfg.map
-            .get_mut(&BlockId::ENTRY)
-            .unwrap()
-            .dominates
-            .extend(blocks.keys().copied().filter(|id| !id.is_entry()));
+        cfg.build_domtree();
+
         cfg
     }
 
+    /// use the Lengauer-Tarjan algorithm to create a dominator tree
+    pub fn build_domtree(&mut self) {
+        // TODO: add a special case that skips all of this when a single-block function is found
+
+        // these can be turned into arrays, but they can be maps for now.
+        // the original algorithm calls for computing predecessors too, but we do that already.
+        #[derive(Debug)]
+        struct LtInfo {
+            parent: Map<BlockId, BlockId>, // pre-order parent
+            semi: Map<BlockId, usize>, // semidominator helper
+            vertex: Map<usize, BlockId>, // block id from pre-order index
+            bucket: Map<BlockId, Set<BlockId>>, // set of blocks whos semidominator is the key
+            forest_parent: Map<BlockId, BlockId>, // intermediate forest used in steps 2 and 3
+            n: usize, // number of blocks
+        }
+        // initialize lengauer-tarjan info
+        let mut lt = LtInfo {
+            parent:        Map::new(),
```

Comment:
This is distinctly un-Rust-like formatting. You can keep this but it'll obviously be erased by rustfmt. Also consider `LtInfo::default()`?